### PR TITLE
[Op] Reduce_Scatter with single tensor as input and shape check

### PR DESCRIPTION
**Pull request by StrongSpoon**
## Description ##
This pr enables op reduce_scatter to receive both single tensor and tensor list as input. Input tensor list is concatenated into a tensor before executing reduce_scatter. Check ensures that input can be equally divided according to rank_list.
In addition, this pr removes useless logic RequestWorkspace.

## Checklist ##

- [x] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

cc @awslabs/raf-reviewer

### CODE REVIEW

File: python/raf/distributed/op.py [hunk 1]
```diff
@@ -151,6 +151,25 @@ def reduce_scatter(x, computation="sum", rank_list=None):
...
+                size = len(group)
+                break
+        else:
+            size = 1
+    else:
+        size = comm.size
+
+    if isinstance(x, (tuple, list)):
+        length = len(x)
+        assert length % size == 0
```
Comment (hgt312) [inline python/raf/distributed/op.py:167]:
```suggestion
        assert length == size
```

**Comment by hgt312** [comment]
@comaniac

**Comment by comaniac** [comment]
Now `reduce_scatter` accepts a tensor instead of a vector of tensors in the C++ side. I just realized that this may break the passes in C++ that makes use of `reduce_scatter`, such as PartitionGradient. For example, should this Tuple be removed https://github.com/awslabs/raf/blob/main/src/pass/partition_gradient.cc#L220?

cc @hgt312 @zhen-jia

**Comment by zhen-jia** [comment]
Yes, that tuple should be removed. @StrongSpoon could you also help to modify the pass accordingly.
> Now `reduce_scatter` accepts a tensor instead of a vector of tensors in the C++ side. I just realized that this may break the passes in C++ that makes use of `reduce_scatter`, such as PartitionGradient. For example, should this Tuple be removed https://github.com/awslabs/raf/blob/main/src/pass/partition_gradient.cc#L220?
> 
> cc @hgt312 @zhen-jia

File: python/raf/distributed/op.py [hunk 2]
```diff
@@ -151,6 +154,31 @@ def reduce_scatter(x, computation="sum", rank_list=None):
...
+                size = len(group)
+                break
+        else:
+            size = 1
+    else:
+        size = comm.size
+
+    if isinstance(x, (tuple, list)):
+        length = len(x)
+        assert length == size
```
Comment (hgt312) [inline python/raf/distributed/op.py:170]:
```suggestion
        assert length == len(x)
```

File: python/raf/distributed/op.py [hunk 3]
```diff
@@ -151,6 +154,31 @@ def reduce_scatter(x, computation="sum", rank_list=None):
...
+        body = Symbol.make_tuple(x)._Symbol__handle
+        body = ExtractBinding(body, [])
+        mod = IRModule.from_expr(body)
+        mod = InferType()(mod)
+        ret_list = mod["main"].checked_type.ret_type
+        single_tensor = ret_list.fields[0]
+        print(type(ret_list.fields))
```
Comment (hgt312) [inline python/raf/distributed/op.py:177]: remove print

**Review by hgt312 — APPROVED** [review]
Otherwise LGTM

File: src/op/declare/collective_comm.cc [hunk 4]
```diff
@@ -124,20 +124,16 @@ RAF_OP_DECLARE("raf.op._group_allgather", GroupAllGather)
...
+  if (args->rank_list.defined()) {
+    size = Communicator::Get("void", args->rank_list)->size;
   } else {
-    for (const auto& tv : tvs) {
-      const DLTensor* x = tv;
-      CHECK(shape == std::vector<int64_t>(x->shape, x->shape + x->ndim));
-    }
+    size = GetGlobalCommunicator()->size;
   }
+  CHECK(shape[0] % size == 0);
```
Comment (comaniac) [inline src/op/declare/collective_comm.cc:135]:
```suggestion
  CHECK(shape[0] % size == 0) << "Input tensor with first dim shape " << shape[0] << " cannot be scattered to " << size << " devices evenly";
```

**Review by comaniac — COMMENTED** [review]
Otherwise LGTM

**Comment by comaniac** [comment]
Thanks @StrongSpoon @hgt312 @zhen-jia

**Pull request merged**